File: laravel-backend/app/DataSources/TechnicalAnalysis.php
<?php

namespace App\DataSources;

/**
 * Technical analysis indicators computed from OHLCV data.
 */
class TechnicalAnalysis
{
    private YahooFinance $yahooFinance;

    public function __construct(YahooFinance $yahooFinance)
    {
        $this->yahooFinance = $yahooFinance;
    }

    public function computeAllTechnicals(array $tickers): array
    {
        $results = [];
        foreach ($tickers as $ticker) {
            $history = $this->yahooFinance->fetchHistory($ticker, '1y');
            if (count($history) < 30) {
                $results[$ticker] = ['error' => "Insufficient history for {$ticker} (" . count($history) . ' days)'];
                continue;
            }
            try {
                $results[$ticker] = $this->computeIndicators($history, $ticker);
            } catch (\Throwable $e) {
                $results[$ticker] = ['error' => $e->getMessage()];
            }
        }
        return $results;
    }

    private function computeIndicators(array $history, string $ticker): array
    {
        $closes = array_map(fn($h) => (float) $h['close'], $history);
        $highs = array_map(fn($h) => (float) $h['high'], $history);
        $lows = array_map(fn($h) => (float) $h['low'], $history);
        $volumes = array_map(fn($h) => (float) $h['volume'], $history);

        $rsi = $this->computeRsi($closes);
        $macd = $this->computeMacd($closes);
        $bollinger = $this->computeBollingerBands($closes);
        $mas = $this->computeMovingAverages($closes);
        $sr = $this->computeSupportResistance($highs, $lows, $closes);
        $vol = $this->computeVolumeAnalysis($volumes);

        $indicators = [
            'ticker' => $ticker,
            'rsi' => $rsi,
            'macd' => $macd,
            'bollinger' => $bollinger,
            'moving_averages' => $mas,
            'support_resistance' => $sr,
            'volume' => $vol,
        ];
        $indicators['overall_signal'] = $this->determineOverallSignal($indicators);

        return $indicators;
    }

    private function computeRsi(array $closes, int $period = 14): array
    {
        $n = count($closes);
        if ($n < $period + 1) {
            return ['value' => null, 'signal' => 'NEUTRAL'];
        }

        $gains = [];
        $losses = [];
        for ($i = 1; $i < $n; $i++) {
            $change = $closes[$i] - $closes[$i - 1];
            $gains[] = $change > 0 ? $change : 0;
            $losses[] = $change < 0 ? abs($change) : 0;
        }

        // Wilder's smoothing (EMA with alpha = 1/period)
        $avgGain = array_sum(array_slice($gains, 0, $period)) / $period;
        $avgLoss = array_sum(array_slice($losses, 0, $period)) / $period;

        for ($i = $period; $i < count($gains); $i++) {
            $avgGain = ($avgGain * ($period - 1) + $gains[$i]) / $period;
            $avgLoss = ($avgLoss * ($period - 1) + $losses[$i]) / $period;
        }

        $rs = $avgLoss > 0 ? $avgGain / $avgLoss : 100;
        $value = round(100 - (100 / (1 + $rs)), 1);

        $signal = 'NEUTRAL';
        if ($value > 70) {
            $signal = 'OVERBOUGHT';
        } elseif ($value < 30) {
            $signal = 'OVERSOLD';
        }

        return ['value' => $value, 'signal' => $signal];
    }

    private function computeMacd(array $closes, int $fast = 12, int $slow = 26, int $signalPeriod = 9): array
    {
        $emaFast = $this->ema($closes, $fast);
        $emaSlow = $this->ema($closes, $slow);

        $n = count($closes);
        $macdLine = [];
        for ($i = 0; $i < $n; $i++) {
            $macdLine[] = $emaFast[$i] - $emaSlow[$i];
        }

        $signalLine = $this->ema($macdLine, $signalPeriod);
        $histogram = [];
        for ($i = 0; $i < $n; $i++) {
            $histogram[] = $macdLine[$i] - $signalLine[$i];
        }

        $macdVal = round(end($macdLine), 4);
        $signalVal = round(end($signalLine), 4);
        $histVal = round(end($histogram), 4);

        $signal = 'NEUTRAL';
        if (count($histogram) >= 2) {
            $prevHist = $histogram[count($histogram) - 2];
            if ($histVal > 0 && $prevHist <= 0) {
                $signal = 'BULLISH';
            } elseif ($histVal < 0 && $prevHist >= 0) {
                $signal = 'BEARISH';
            } elseif ($histVal > 0) {
                $signal = 'BULLISH';
            } else {
                $signal = 'BEARISH';
            }
        }

        return [
            'macd_line' => $macdVal,
            'signal_line' => $signalVal,
            'histogram' => $histVal,
            'signal' => $signal,
        ];
    }

    private function computeBollingerBands(array $closes, int $period = 20, float $stdDev = 2.0): array
    {
        $n = count($closes);
        $slice = array_slice($closes, -$period);
        $sma = array_sum($slice) / $period;
        $std = $this->stdDev($slice);

        $upper = $sma + $stdDev * $std;
        $lower = $sma - $stdDev * $std;
        $price = end($closes);

        $bandWidth = $upper - $lower;
        $position = $bandWidth > 0 ? ($price - $lower) / $bandWidth : 0.5;

        $squeeze = false;
        if ($n >= 120) {
            $currentBandwidth = $bandWidth / $sma;
            // Approximate average bandwidth
            $bws = [];
            for ($i = $period; $i <= $n; $i++) {
                $s = array_slice($closes, $i - $period, $period);
                $m = array_sum($s) / $period;
                $sd = $this->stdDev($s);
                if ($m > 0) {
                    $bws[] = (2 * $stdDev * $sd) / $m;
                }
            }
            $avgBw = ! empty($bws) ? array_sum($bws) / count($bws) : $currentBandwidth;
            $squeeze = $currentBandwidth < $avgBw * 0.75;
        }

        $signal = 'NEUTRAL';
        if ($position > 0.8) {
            $signal = 'OVERBOUGHT';
        } elseif ($position < 0.2) {
            $signal = 'OVERSOLD';
        }

        return [
            'upper' => round($upper, 2),
            'middle' => round($sma, 2),
            'lower' => round($lower, 2),
            'bandwidth' => round($bandWidth / ($sma ?: 1), 4),
            'position' => round($position, 3),
            'squeeze' => $squeeze,
            'signal' => $signal,
        ];
    }

    private function computeMovingAverages(array $closes): array
    {
        $n = count($closes);
        $price = end($closes);

        $sma20 = $n >= 20 ? round(array_sum(array_slice($closes, -20)) / 20, 2) : null;
        $sma50 = $n >= 50 ? round(array_sum(array_slice($closes, -50)) / 50, 2) : null;
        $sma200 = $n >= 200 ? round(array_sum(array_slice($closes, -200)) / 200, 2) : null;

        $goldenCross = false;
        $deathCross = false;

        if ($sma50 !== null && $sma200 !== null && $n >= 201) {
            $prevSma50 = array_sum(array_slice($closes, -51, 50)) / 50;
            $prevSma200 = array_sum(array_slice($closes, -201, 200)) / 200;
            $diffNow = $sma50 - $sma200;
            $diffPrev = $prevSma50 - $prevSma200;
            if ($diffNow > 0 && $diffPrev <= 0) {
                $goldenCross = true;
            } elseif ($diffNow < 0 && $diffPrev >= 0) {
                $deathCross = true;
            }
        }

        $priceVsSma200 = 'N/A';
        if ($sma200 !== null) {
            $priceVsSma200 = $price > $sma200 ? 'ABOVE' : 'BELOW';
        }

        return [
            'sma_20' => $sma20,
            'sma_50' => $sma50,
            'sma_200' => $sma200,
            'golden_cross' => $goldenCross,
            'death_cross' => $deathCross,
            'price_vs_sma200' => $priceVsSma200,
        ];
    }

    private function computeSupportResistance(array $highs, array $lows, array $closes, int $lookback = 60): array
    {
        $price = end($closes);
        $n = count($highs);
        $start = max(0, $n - $lookback);

        $recentHighs = array_slice($highs, $start);
        $recentLows = array_slice($lows, $start);
        $window = 5;

        $resistanceLevels = [];
        $supportLevels = [];

        $rLen = count($recentHighs);
        for ($i = $window; $i < $rLen - $window; $i++) {
            $localMax = max(array_slice($recentHighs, $i - $window, $window * 2 + 1));
            if ($recentHighs[$i] == $localMax) {
                $resistanceLevels[] = $recentHighs[$i];
            }
            $localMin = min(array_slice($recentLows, $i - $window, $window * 2 + 1));
            if ($recentLows[$i] == $localMin) {
                $supportLevels[] = $recentLows[$i];
            }
        }

        $supportsBelow = array_filter($supportLevels, fn($s) => $s < $price);
        $resistancesAbove = array_filter($resistanceLevels, fn($r) => $r > $price);

        $nearestSupport = ! empty($supportsBelow) ? max($supportsBelow) : min($recentLows);
        $nearestResistance = ! empty($resistancesAbove) ? min($resistancesAbove) : max($recentHighs);

        return [
            'nearest_support' => round($nearestSupport, 2),
            'nearest_resistance' => round($nearestResistance, 2),
            'support_distance_pct' => $price > 0 ? round((($price - $nearestSupport) / $price) * -100, 2) : 0,
            'resistance_distance_pct' => $price > 0 ? round((($nearestResistance - $price) / $price) * 100, 2) : 0,
        ];
    }

    private function computeVolumeAnalysis(array $volumes): array
    {
        $current = end($volumes);
        $n = count($volumes);
        $avg20d = $n >= 20
            ? array_sum(array_slice($volumes, -20)) / 20
            : $current;

        $ratio = $avg20d > 0 ? round($current / $avg20d, 2) : 1.0;

        $signal = 'NORMAL';
        if ($ratio > 1.5) {
            $signal = 'HIGH';
        } elseif ($ratio < 0.5) {
            $signal = 'LOW';
        }

        return [
            'current' => (int) $current,
            'avg_20d' => (int) $avg20d,
            'ratio' => $ratio,
            'signal' => $signal,
        ];
    }

    private function determineOverallSignal(array $indicators): string
    {
        $score = 0;

        $rsiSignal = $indicators['rsi']['signal'] ?? 'NEUTRAL';
        if ($rsiSignal === 'OVERSOLD') $score++;
        elseif ($rsiSignal === 'OVERBOUGHT') $score--;

        $macdSignal = $indicators['macd']['signal'] ?? 'NEUTRAL';
        if ($macdSignal === 'BULLISH') $score++;
        elseif ($macdSignal === 'BEARISH') $score--;

        $smaPos = $indicators['moving_averages']['price_vs_sma200'] ?? 'N/A';
        if ($smaPos === 'ABOVE') $score++;
        elseif ($smaPos === 'BELOW') $score--;

        $bbPos = $indicators['bollinger']['position'] ?? 0.5;
        if ($bbPos < 0.2) $score++;
        elseif ($bbPos > 0.8) $score--;

        if ($indicators['moving_averages']['golden_cross'] ?? false) $score++;
        if ($indicators['moving_averages']['death_cross'] ?? false) $score--;

        if ($score >= 2) return 'BULLISH';
        if ($score <= -2) return 'BEARISH';
        return 'NEUTRAL';
    }

    private function ema(array $data, int $period): array
    {
        $n = count($data);
        $result = array_fill(0, $n, 0.0);
        $multiplier = 2.0 / ($period + 1);

        // Initialize with SMA
        $result[0] = $data[0];
        for ($i = 1; $i < $n; $i++) {
            $result[$i] = ($data[$i] - $result[$i - 1]) * $multiplier + $result[$i - 1];
        }

        return $result;
    }

    private function stdDev(array $data): float
    {
        $n = count($data);
        if ($n < 2) return 0;
        $mean = array_sum($data) / $n;
        $sumSquares = array_sum(array_map(fn($x) => ($x - $mean) ** 2, $data));
        return sqrt($sumSquares / ($n - 1));
    }
}

File: backend/app/data_sources/technical_analysis.py
"""Technical analysis indicators computed from OHLCV data via pandas/numpy."""

import numpy as np
import pandas as pd
from app.data_sources import massive


def compute_all_technicals(tickers: list[str]) -> dict:
    """Compute technical indicators for all tickers."""
    results = {}
    for ticker in tickers:
        history = massive.fetch_history(ticker, period="1y")
        if len(history) < 30:
            results[ticker] = {"error": f"Insufficient history for {ticker} ({len(history)} days)"}
            continue
        try:
            results[ticker] = _compute_indicators(history, ticker)
        except Exception as e:
            results[ticker] = {"error": str(e)}
    return results


def _compute_indicators(history: list[dict], ticker: str) -> dict:
    """Compute all indicators for a single ticker from OHLCV history."""
    df = pd.DataFrame(history)
    closes = df["close"].astype(float)
    highs = df["high"].astype(float)
    lows = df["low"].astype(float)
    volumes = df["volume"].astype(float)

    rsi = _compute_rsi(closes)
    macd = _compute_macd(closes)
    bollinger = _compute_bollinger_bands(closes)
    mas = _compute_moving_averages(closes)
    sr = _compute_support_resistance(highs, lows, closes)
    vol = _compute_volume_analysis(volumes)

    indicators = {
        "ticker": ticker,
        "rsi": rsi,
        "macd": macd,
        "bollinger": bollinger,
        "moving_averages": mas,
        "support_resistance": sr,
        "volume": vol,
    }
    indicators["overall_signal"] = _determine_overall_signal(indicators)
    return indicators


def _compute_rsi(closes: pd.Series, period: int = 14) -> dict:
    """RSI via Wilder's smoothing."""
    delta = closes.diff()
    gain = delta.where(delta > 0, 0.0)
    loss = -delta.where(delta < 0, 0.0)
    avg_gain = gain.ewm(alpha=1 / period, min_periods=period).mean()
    avg_loss = loss.ewm(alpha=1 / period, min_periods=period).mean()
    rs = avg_gain / avg_loss
    rsi = 100 - (100 / (1 + rs))
    value = round(float(rsi.iloc[-1]), 1) if not rsi.empty else None

    if value is None:
        signal = "NEUTRAL"
    elif value > 70:
        signal = "OVERBOUGHT"
    elif value < 30:
        signal = "OVERSOLD"
    else:
        signal = "NEUTRAL"

    return {"value": value, "signal": signal}


def _compute_macd(
    closes: pd.Series, fast: int = 12, slow: int = 26, signal_period: int = 9
) -> dict:
    """MACD line, signal line, histogram."""
    ema_fast = closes.ewm(span=fast, adjust=False).mean()
    ema_slow = closes.ewm(span=slow, adjust=False).mean()
    macd_line = ema_fast - ema_slow
    signal_line = macd_line.ewm(span=signal_period, adjust=False).mean()
    histogram = macd_line - signal_line

    macd_val = round(float(macd_line.iloc[-1]), 4)
    signal_val = round(float(signal_line.iloc[-1]), 4)
    hist_val = round(float(histogram.iloc[-1]), 4)

    # Crossover detection
    if len(histogram) >= 2:
        prev_hist = float(histogram.iloc[-2])
        if hist_val > 0 and prev_hist <= 0:
            signal = "BULLISH"
        elif hist_val < 0 and prev_hist >= 0:
            signal = "BEARISH"
        elif hist_val > 0:
            signal = "BULLISH"
        else:
            signal = "BEARISH"
    else:
        signal = "NEUTRAL"

    return {
        "macd_line": macd_val,
        "signal_line": signal_val,
        "histogram": hist_val,
        "signal": signal,
    }


def _compute_bollinger_bands(
    closes: pd.Series, period: int = 20, std_dev: float = 2.0
) -> dict:
    """Bollinger Bands with squeeze detection."""
    sma = closes.rolling(period).mean()
    std = closes.rolling(period).std()
    upper = sma + std_dev * std
    lower = sma - std_dev * std

    upper_val = float(upper.iloc[-1])
    lower_val = float(lower.iloc[-1])
    middle_val = float(sma.iloc[-1])
    price = float(closes.iloc[-1])

    band_width = upper_val - lower_val
    position = (price - lower_val) / band_width if band_width > 0 else 0.5

    # Squeeze: current bandwidth below 6-month average bandwidth
    bandwidth_series = (upper - lower) / sma
    avg_bandwidth = float(bandwidth_series.rolling(120).mean().iloc[-1]) if len(bandwidth_series) >= 120 else float(bandwidth_series.mean())
    current_bandwidth = float(bandwidth_series.iloc[-1])
    squeeze = current_bandwidth < avg_bandwidth * 0.75

    if position > 0.8:
        signal = "OVERBOUGHT"
    elif position < 0.2:
        signal = "OVERSOLD"
    else:
        signal = "NEUTRAL"

    return {
        "upper": round(upper_val, 2),
        "middle": round(middle_val, 2),
        "lower": round(lower_val, 2),
        "bandwidth": round(current_bandwidth, 4),
        "position": round(position, 3),
        "squeeze": squeeze,
        "signal": signal,
    }


def _compute_moving_averages(closes: pd.Series) -> dict:
    """SMA 20/50/200 with golden/death cross detection."""
    price = float(closes.iloc[-1])
    sma_20 = float(closes.rolling(20).mean().iloc[-1]) if len(closes) >= 20 else None
    sma_50 = float(closes.rolling(50).mean().iloc[-1]) if len(closes) >= 50 else None
    sma_200 = float(closes.rolling(200).mean().iloc[-1]) if len(closes) >= 200 else None

    # Golden/death cross: SMA50 crossing SMA200
    golden_cross = False
    death_cross = False
    if sma_50 is not None and sma_200 is not None and len(closes) >= 201:
        sma50_series = closes.rolling(50).mean()
        sma200_series = closes.rolling(200).mean()
        diff_now = sma50_series.iloc[-1] - sma200_series.iloc[-1]
        diff_prev = sma50_series.iloc[-2] - sma200_series.iloc[-2]
        if diff_now > 0 and diff_prev <= 0:
            golden_cross = True
        elif diff_now < 0 and diff_prev >= 0:
            death_cross = True

    price_vs_sma200 = "ABOVE" if sma_200 and price > sma_200 else "BELOW" if sma_200 else "N/A"

    return {
        "sma_20": round(sma_20, 2) if sma_20 else None,
        "sma_50": round(sma_50, 2) if sma_50 else None,
        "sma_200": round(sma_200, 2) if sma_200 else None,
        "golden_cross": golden_cross,
        "death_cross": death_cross,
        "price_vs_sma200": price_vs_sma200,
    }


def _compute_support_resistance(
    highs: pd.Series, lows: pd.Series, closes: pd.Series, lookback: int = 60
) -> dict:
    """Identify support/resistance from recent price pivots."""
    price = float(closes.iloc[-1])
    recent_highs = highs.tail(lookback)
    recent_lows = lows.tail(lookback)

    # Find local maxima and minima using rolling windows
    window = 5
    resistance_levels = []
    support_levels = []

    for i in range(window, len(recent_highs) - window):
        if recent_highs.iloc[i] == recent_highs.iloc[i - window : i + window + 1].max():
            resistance_levels.append(float(recent_highs.iloc[i]))
        if recent_lows.iloc[i] == recent_lows.iloc[i - window : i + window + 1].min():
            support_levels.append(float(recent_lows.iloc[i]))

    # Find nearest support below price and resistance above price
    supports_below = [s for s in support_levels if s < price]
    resistances_above = [r for r in resistance_levels if r > price]

    nearest_support = max(supports_below) if supports_below else float(recent_lows.min())
    nearest_resistance = min(resistances_above) if resistances_above else float(recent_highs.max())

    support_dist = round(((price - nearest_support) / price) * -100, 2)
    resistance_dist = round(((nearest_resistance - price) / price) * 100, 2)

    return {
        "nearest_support": round(nearest_support, 2),
        "nearest_resistance": round(nearest_resistance, 2),
        "support_distance_pct": support_dist,
        "resistance_distance_pct": resistance_dist,
    }


def _compute_volume_analysis(volumes: pd.Series) -> dict:
    """Current volume vs 20-day average."""
    current = float(volumes.iloc[-1])
    avg_20d = float(volumes.rolling(20).mean().iloc[-1]) if len(volumes) >= 20 else current
    ratio = round(current / avg_20d, 2) if avg_20d > 0 else 1.0

    if ratio > 1.5:
        signal = "HIGH"
    elif ratio < 0.5:
        signal = "LOW"
    else:
        signal = "NORMAL"

    return {
        "current": int(current),
        "avg_20d": int(avg_20d),
        "ratio": ratio,
        "signal": signal,
    }


def _determine_overall_signal(indicators: dict) -> str:
    """Score-based aggregate signal."""
    score = 0

    # RSI
    rsi_signal = indicators.get("rsi", {}).get("signal", "NEUTRAL")
    if rsi_signal == "OVERSOLD":
        score += 1
    elif rsi_signal == "OVERBOUGHT":
        score -= 1

    # MACD
    macd_signal = indicators.get("macd", {}).get("signal", "NEUTRAL")
    if macd_signal == "BULLISH":
        score += 1
    elif macd_signal == "BEARISH":
        score -= 1

    # Price vs SMA200
    sma_pos = indicators.get("moving_averages", {}).get("price_vs_sma200", "N/A")
    if sma_pos == "ABOVE":
        score += 1
    elif sma_pos == "BELOW":
        score -= 1

    # Bollinger position
    bb_pos = indicators.get("bollinger", {}).get("position", 0.5)
    if bb_pos < 0.2:
        score += 1
    elif bb_pos > 0.8:
        score -= 1

    # Golden/Death cross
    if indicators.get("moving_averages", {}).get("golden_cross"):
        score += 1
    if indicators.get("moving_averages", {}).get("death_cross"):
        score -= 1

    if score >= 2:
        return "BULLISH"
    elif score <= -2:
        return "BEARISH"
    return "NEUTRAL"
